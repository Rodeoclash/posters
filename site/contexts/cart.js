import { createContext, useContext } from "react";

const CartContext = createContext({
  addProduct: () => {},
  inCart: () => {},
  removeProduct: () => {},
  size: 0,
});

export const useCartContext = () => useContext(CartContext);

export default CartContext;
