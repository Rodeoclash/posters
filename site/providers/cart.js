import { useState } from "react";
import CartContext from "../contexts/cart.js";

const LoadingProvider = ({ children }) => {
  const [state, setState] = useState({
    products: [],
  });

  const addProduct = (product) => {
    setState({
      ...state,
      products: state.products.concat([product]),
    });
  };

  const value = {
    addProduct,
    products: state.products,
    size: state.products.length,
  };

  return <CartContext.Provider value={value}>{children}</CartContext.Provider>;
};

export default LoadingProvider;
