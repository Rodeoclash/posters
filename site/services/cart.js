export class ProductNotFound extends Error {}
export class ProductExists extends Error {}

const productMatch = (product1, product2) => {
  return product1.id === product2.id;
};

const productNoMatch = (product1, product2) => {
  return !productMatch(product1, product2);
};

export const show = (cart) => {
  return {
    ...cart,
    showing: true,
  };
};

export const hide = (cart) => {
  return {
    ...cart,
    showing: false,
  };
};

export const contains = (cart, product) => {
  return cart.products.some(productMatch.bind(this, product));
};

export const add = (cart, product) => {
  if (contains(cart, product) === true) {
    throw new ProductExists();
  }

  return {
    ...show(cart),
    products: cart.products.concat([product]),
  };
};

export const remove = (cart, product) => {
  if (contains(cart, product) === false) {
    throw new ProductNotFound();
  }

  return {
    ...show(cart),
    products: cart.products.filter(productNoMatch.bind(this, product)),
  };
};
