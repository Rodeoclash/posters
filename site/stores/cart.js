import { createContext } from "react";
import { useObservable } from "mobx-react-lite";

const CartContext = createContext();

function Cart({ children }) {
  const store = useObservable({ counter: 0 });
  return <CartContext.Provider value={store}>{children}</CartContext.Provider>;
}

export { Cart, CartContext };
