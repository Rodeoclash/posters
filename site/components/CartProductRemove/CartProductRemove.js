import { useCartContext } from "../../contexts/cart";

import styles from "./CartProductRemove.module.css";

const CartProductRemove = ({ product }) => {
  const { removeProduct, productInCart } = useCartContext();

  if (productInCart(product) === false) {
    return null;
  }

  return (
    <button onClick={removeProduct.bind(this, product)}>Remove product</button>
  );
};

export default CartProductRemove;
