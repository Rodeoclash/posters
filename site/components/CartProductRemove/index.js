import CartProductRemove from "./CartProductRemove";
export default CartProductRemove;
