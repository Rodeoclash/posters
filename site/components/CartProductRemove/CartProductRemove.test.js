import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render } from "@testing-library/react";

import * as CartContext from "../../contexts/cart";
import CartProductRemove from "./CartProductRemove";

describe("onClick", () => {
  it("calls function to remove product from cart", () => {
    const product = {};
    const remove = jest.fn();

    const contextValues = {
      contains() {
        return true;
      },
      remove,
    };

    jest
      .spyOn(CartContext, "useCartContext")
      .mockImplementation(() => contextValues);

    const { getByText } = render(<CartProductRemove product={product} />);

    fireEvent.click(getByText("Remove product"));

    expect(remove).toHaveBeenCalledWith(product, expect.anything());
  });
});

describe("snapshots", () => {
  it("with product in the cart", () => {
    const product = {};
    const remove = jest.fn();

    const contextValues = {
      contains() {
        return true;
      },
      remove,
    };

    jest
      .spyOn(CartContext, "useCartContext")
      .mockImplementation(() => contextValues);

    const tree = renderer
      .create(<CartProductRemove product={product} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("with product not in the cart", () => {
    const product = {};
    const remove = jest.fn();

    const contextValues = {
      contains() {
        return false;
      },
      remove,
    };

    jest
      .spyOn(CartContext, "useCartContext")
      .mockImplementation(() => contextValues);

    const tree = renderer
      .create(<CartProductRemove product={product} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
