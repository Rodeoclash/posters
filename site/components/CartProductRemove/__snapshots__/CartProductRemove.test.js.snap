// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots with product in the cart 1`] = `
<button
  onClick={[Function]}
>
  Remove product
</button>
`;

exports[`snapshots with product not in the cart 1`] = `null`;
